quadtree/src/lib.rs

```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
     }
 }
 
+pub enum SpatialNeighbor {
+    Left,
+    TopLeft,
+    Top,
+    TopRight,
+    Right,
+    BottomRight,
+    Bottom,
+    BottomLeft,
+}
+
+#[derive(Clone, Copy)]
+pub struct SpatialNodeId {
+    level: u8,
+    x: u64,
+    y: u64,
+}
+
+impl SpatialNodeId {
+    pub fn new(level: u8, x: u64, y: u64) -> Self {
+        Self { level, x, y }
+    }
+
+    pub fn neighbor(&self, kind: SpatialNeighbor) -> Option<Self> {
+        let cur_x = self.x as i64;
+        let cur_y = self.y as i64;
+        let (x, y) = match kind {
+            SpatialNeighbor::Left => (cur_x - 1, cur_y),
+            SpatialNeighbor::TopLeft => (cur_x - 1, cur_y + 1),
+            SpatialNeighbor::Top => (cur_x, cur_y + 1),
+            SpatialNeighbor::TopRight => (cur_x + 1, cur_y + 1),
+            SpatialNeighbor::Right => (cur_x + 1, cur_y),
+            SpatialNeighbor::BottomRight => (cur_x + 1, cur_y - 1),
+            SpatialNeighbor::Bottom => (cur_x, cur_y - 1),
+            SpatialNeighbor::BottomLeft => (cur_x - 1, cur_y - 1),
+        };
+        let max_dim = 2i64.pow(self.level as u32);
+        if 0 <= x && x < max_dim && 0 <= y && y < max_dim {
+            Some(Self::new(self.level, x as u64, y as u64))
+        } else {
+            None
+        }
+    }
+}
+
+/// See e.g. https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
+/// on how to convert between coordinates and the quadkey.
+impl From<NodeId> for SpatialNodeId {
+    fn from(node_id: NodeId) -> Self {
+        let level = node_id.level;
+        let mut x = 0;
+        let mut y = 0;
+        for i in 1..=level {
+            let mask = 1 << (level - i);
+            match 3 & (node_id.index >> ((level - i) * 2)) {
+                1 => y |= mask,
```

Comment:
This is different than what's in the link and your coordinates are transposed w.r.t. the coordinates in the doc, in the sense that whenever they have `y` you have `x`, and whenever they have `x` you have `y` (I hope this makes sense).
Is it intentional?

Reply:
Actually we here have a right hand coordinate frame (in contrast to the left hand frame for the Bing link), so I had to switch x and y.

Reply:
:+1: